### PR TITLE
Extract DIMACS Printing

### PR DESCRIPTION
**Pull request by alex-ozdemir**
Creating LRAT proofs requires writing SAT problems in the DIMACS format.
Before this code was in the LRAT class.

However, since creating ER proofs will also require writing DIMACS, I
decided to extract it this code into helper functions.

At the same time I realized that my prior representation of used clauses
was unnecessarily poor. I had chosen it to interface with
`CnfProof::collectAtomsForClauses`, but I have since discovered that the
aforementioned method is super simple, so choosing a poor representation to re-use it is not the right call.

### CODE REVIEW

File: src/proof/dimacs_printer.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include <memory>
+
+#include "proof/clause_id.h"
+#include "prop/sat_solver_types.h"
+
+namespace CVC4 {
+namespace proof {
+
+// Prints the literal as a (+) or (-) int
+// Not operator<< b/c that represents negation as ~
```
Comment (aniemetz) [inline src/proof/dimacs_printer.h]: Use a style consistent with comments in other header files:
```
/* ... */

/**
  ...
  ...
*/
```

File: src/proof/dimacs_printer.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+std::ostream& textOut(std::ostream& o, const prop::SatLiteral& l);
+
+// Prints the clause as a space-separated list of ints
+// Not operator<< b/c that represents negation as ~
```
Comment (aniemetz) [inline src/proof/dimacs_printer.h]: Ditto

File: src/proof/dimacs_printer.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+std::ostream& textOut(std::ostream& o, const prop::SatClause& c);
+
+void printDimacs(
```
Comment (aniemetz) [inline src/proof/dimacs_printer.h]: Document.

File: src/proof/dimacs_printer.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    o << "-";
+  }
+  return o << l.getSatVariable() + 1;
+}
+
+// Prints the clause as a space-separated list of ints
+// Not operator<< b/c that represents negation as ~
+std::ostream& textOut(std::ostream& o, const prop::SatClause& c)
+{
+  for (const auto l : c)
```
Comment (aniemetz) [inline src/proof/dimacs_printer.cpp]: `for (const auto& ...)`

File: src/proof/dimacs_printer.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+void printDimacs(
+    std::ostream& o,
+    const std::vector<std::pair<ClauseId, prop::SatClause>>& usedClauses)
+{
+  size_t maxVar = 0;
+  for (auto& c : usedClauses)
+  {
+    for (auto l : c.second)
```
Comment (aniemetz) [inline src/proof/dimacs_printer.cpp]: `for (const auto& ...)`

File: src/proof/dimacs_printer.cpp (hunk 4 above)
Comment (alex-ozdemir) [reply src/proof/dimacs_printer.cpp]: Sounds good!

File: src/proof/dimacs_printer.cpp (hunk 5 above)
Comment (alex-ozdemir) [reply src/proof/dimacs_printer.cpp]: Sounds good!

File: src/proof/dimacs_printer.h (hunk 1 above)
Comment (alex-ozdemir) [reply src/proof/dimacs_printer.h]: Sounds good!

File: src/proof/dimacs_printer.h (hunk 2 above)
Comment (alex-ozdemir) [reply src/proof/dimacs_printer.h]: :+1:

File: src/proof/dimacs_printer.h (hunk 3 above)
Comment (alex-ozdemir) [reply src/proof/dimacs_printer.h]: Whoops! Good catch. Will do!

**Comment by alex-ozdemir** [comment]
Thanks for the review, Aina! I believe that I've addressed each of your comments. Let me know if I missed any, or if further work should be done.

**Pull request merged**